checkov/terraform/module_loading/loaders/git_loader.py

```diff
@@ -1,31 +1,103 @@
 import os
+import re
+from dataclasses import dataclass
+from pathlib import Path
 
 from checkov.common.goget.github.get_git import GitGetter
 from checkov.terraform.module_loading.content import ModuleContent
 
 from checkov.terraform.module_loading.loader import ModuleLoader
 
 
+@dataclass(frozen=True)
+class ModuleSource:
+    protocol: str
+    root_module: str
+    inner_module: str
+    version: str
+
+
 class GenericGitLoader(ModuleLoader):
-    def _is_matching_loader(self):
+    def _is_matching_loader(self) -> bool:
         # https://www.terraform.io/docs/modules/sources.html#generic-git-repository
-        return self.module_source.startswith('git::')
+        return self.module_source.startswith("git::")
 
     def _load_module(self) -> ModuleContent:
         try:
-            module_source = self.module_source.replace('git::', '')
-            if module_source.startswith('ssh:'):
-                return ModuleContent(dir=None, failed_url=self.module_source)
+            self._process_generic_git_repo()
+
+            module_source = self.module_source.replace("git::", "")
             git_getter = GitGetter(module_source, create_clone_and_result_dirs=False)
             git_getter.temp_dir = self.dest_dir
             git_getter.do_get()
+
             return_dir = self.dest_dir
             if self.inner_module:
                 return_dir = os.path.join(self.dest_dir, self.inner_module)
             return ModuleContent(dir=return_dir)
         except Exception as e:
-            self.logger.error(f'failed to get {self.module_source} because of {e}')
+            self.logger.error(f"failed to get {self.module_source} because of {e}")
             return ModuleContent(dir=None, failed_url=self.module_source)
 
+    def _find_module_path(self) -> str:
+        module_source = self._parse_module_source()
+        module_path = Path(self.current_dir).joinpath(
+            self.external_modules_folder_name,
+            module_source.root_module,
+            module_source.version,
+            module_source.inner_module,
+        )
+
+        return str(module_path)
+
+    def _parse_module_source(self) -> ModuleSource:
+        module_source_components = self.module_source.split("//")
+
+        if "?ref=" in module_source_components[-1]:
+            module_version_components = module_source_components[-1].rsplit("?ref=", maxsplit=1)
+            module_source_components[-1] = module_version_components[0]
+            version = module_version_components[1]
+        else:
+            version = "HEAD"
+
+        if len(module_source_components) == 2:
+            inner_module = ""
+        elif len(module_source_components) == 3:
+            inner_module = module_source_components[2]
+        else:
+            raise Exception("invalid git url")
```

Comment:
@gruebel In general, in this area (module loading) we prefer logging warnings and skipping rather than raising exceptions and failing, Is this caught anywhere?

Reply:
yeah, it is caught in the `parser.py` with an appropriate message
https://github.com/bridgecrewio/checkov/blob/0f63f8e5be52d2a4b9c1b46d7952eddb717e673c/checkov/terraform/parser.py#L396-L397
https://github.com/bridgecrewio/checkov/blob/0f63f8e5be52d2a4b9c1b46d7952eddb717e673c/checkov/terraform/parser.py#L453-L455